### PR TITLE
Fix #176 : Create new Service Proxy or regenerate existing Service Proxy fails (OData V3, WCF Data Services)

### LINKED ISSUE #176: Create new Service Proxy or regenerate existing Service Proxy fails (OData V3, WCF Data Services)
Generating a new client proxy fails with "GetMSBuildProjectForHierarchy must be called on the UI thread".

ODataConnectedService v0.11.1
VisualStudio Version v16.7.7
Microsoft.Data.Odata v5.8.4
Microsoft.Data.Services.Client v5.8.4
Microsoft.Data.Services v5.8.4

1. Create WebApi project with simple code-first EF context and WCF Data Service for this context.
2. Create MVC project for client
3. Browse the Data Service Url (VS show in browser)
4. In the client project add connected service / ODataConnectedService
5. Set Address to the service Url
6. Click Finish

The progess dialog shows the error:
image

Clicking on the Next button after step 5 (to Schema Types Selection) the UI stops reacting on the wizard buttons except the Cancel button.
Clicking on Cancel button or closing the wizard window closes the window.
Clicking on the Finish button first and then on the Cancel button or closing the wizard window results in the following error:
image
- looks like Dictionary Get with NULL key...

This is the same behavior when trying to regenerate an existing service proxy:
1. Select Refresh Connected Service on existing service
2. Click finish in the Wizard window

Results in the error in the last screenshot. 
The behavior after clicking the Next button ist the same as decribed in the case of a new service proxy.  

Update: 
ODataConnectedService v0.10.0 works with one exception:
Refreshing an existing service proxy: clicking on Finish without first clicking Next causes the same error.

### PR DESCRIPTION
**Pull request by unchase**
Fix #176.
There are a few errors in v0.11.1:

issue176

issue176_1

issue176_2

for metadata:

```xml
<?xml version="1.0" standalone="yes" ?>
<edmx:Edmx Version="1.0" xmlns:edmx="http://schemas.microsoft.com/ado/2006/04/edmx">
  <edmx:DataServices xmlns:m="http://schemas.microsoft.com/ado/2007/08/dataservices/metadata" m:DataServiceVersion="2.0">
    <Schema Namespace="Simple" xmlns="http://schemas.microsoft.com/ado/2006/04/edmx">
      <EntityType Name="TestType">
        <Key>
          <PropertyRef Name="KeyProp" />
        </Key>
        <Property Name="KeyProp" Type="Edm.Int32" Nullable="false" />
        <Property Name="ValueProp" Type="Edm.String" Nullable="false" />
      </EntityType>
    </Schema>
  </edmx:DataServices>
</edmx:Edmx> 
```

fixed.

---

Added the tests below:

To **EdmHelperTests:**

- [x] `TestGetSchemaTypesWithNullModel`;
- [x] `TestGetOperationImportsWithNullModel`;
- [x] `TestGetBoundOperationsWithNullModel`.

To **CodeGenDescriptorTest:**

- [x] `TestV2AddGeneratedClientCode_GeneratesCodeForv2_ForVB`;
- [x] `TestV2AddGeneratedClientCode_GeneratesCodeForv2`.

---

Also improved logging the *'Could not get CLR type name for EDM type'* exception in OData v4 code generation from:

edmTypeError

to:

edmTypeErrorDetalization

for metadata (for example): 

```xml
<?xml version="1.0" standalone="yes" ?>
<edmx:Edmx Version="4.0" xmlns:edmx="http://docs.oasis-open.org/odata/ns/edmx">
  <edmx:DataServices>
    <Schema Namespace="Simple" xmlns:d="http://docs.oasis-open.org/odata/ns/data" 
            xmlns:m="http://docs.oasis-open.org/odata/ns/metadata" 
            xmlns="http://docs.oasis-open.org/odata/ns/edm">
      <EntityType Name="TestType">
        <Key>
          <PropertyRef Name="KeyProp" />
        </Key>
        <Property Name="KeyProp" Type="Edm.Int32" Nullable="false" />
        <Property Name="ValueProp" Type="Edm.String" Nullable="false" />
      </EntityType>
	  <EntityContainer Name="Container">
		<ActionImport Name="Simple.NotExistedAction" Action="SomeNamespace.NotExistedAction" />
	  </EntityContainer>
    </Schema>
  </edmx:DataServices>
</edmx:Edmx>
```

### CODE REVIEW

**Comment by unchase** [comment]
@habbes @marabooy @KenitoInc @ElizabethOkerio @gathogojr @paulodero ready for review

There is a false positive error:
image

How can I fix it?

**Review by marabooy — APPROVED** [review]

**Review by KenitoInc — CHANGES_REQUESTED** [review]
We can fix the ".......must be called on the UI thread" errors in V3 the way I did in V4.
Compare the `V3CodeGenDescriptor.cs` and `V4CodeGenDescriptor.cs`

File: src/CodeGeneration/V4CodeGenDescriptor.cs [hunk 1]
```diff
@@ -66,7 +66,7 @@ private async Task AddT4FileAsync()
             var tempFile = Path.GetTempFileName();
             var t4Folder = Path.Combine(this.CurrentAssemblyPath, "Templates");
 
-            var referenceFolder = GetReferenceFileFolder();
+            var referenceFolder = this.GetReferenceFileFolder();
```
Comment (KenitoInc) [inline src/CodeGeneration/V4CodeGenDescriptor.cs:69]: Why are you adding `this.` ?

Comment (unchase) [reply src/CodeGeneration/V4CodeGenDescriptor.cs:69]: image

**Review by KenitoInc — APPROVED** [review]
LGTM

**Review by gathogojr — APPROVED** [review]
LGTM

**Pull request merged**